### PR TITLE
#170947381 Users should be able to Login

### PR DESCRIPTION
**Pull request by blaise82**
#### What does this PR do?
implements a login form and signing page

#### Description of Task to be completed?
- create a login component
- add form validations
- update theme
- update Webpack

#### How should this be manually tested?

- Git clone this repo and checkout this branch `git checkout ft-user-signin-170947381`
- Run npm install
- Run npm dev to start client-server in development mood
- login in with these correct credentials
```
email : coming@soon.com
password: password
```
- Run npm run storybook:dev to view components in the storybook

#### What are the relevant pivotal tracker stories?
https://www.pivotaltracker.com/story/show/170947381

#### Screenshots (if appropriate)
Desktop View 
screencapture-localhost-8080-2020-04-22-00_08_03

[finishes #170947381]

### CODE REVIEW

**Comment by Moise1** [comment]
@blaise82 , I can see that you direct a reviewer to checkout "this branch" but you didn't provide that branch name in the PR's description. Please double-check.

File: src/App.js [hunk 1]
```diff
@@ -1,32 +1,25 @@
...
-	  );
-  }
+			<Provider store={store}>
+				<Router>
+					<div className="center">
+						<Switch>
+							<Route path="/" exact component={SignIn} />
+						</Switch>
+					</div>
+				</Router>
```
Comment (AJAkimana) [inline src/App.js]: Create a file where you will put all routes. Routes don't have to be in this file

File: .babelrc [hunk 2]
```diff
@@ -2,5 +2,10 @@
...
         "@babel/preset-env",
         "@babel/preset-react"
+    ],
+    "plugins": [
+        [
+            "@babel/plugin-proposal-class-properties"
+        ]
     ]
-}
+}
```
Comment (AJAkimana) [inline .babelrc:10]: Add a new line

**Review by AJAkimana — COMMENTED** [review]
Add `package-lock` file to gitignore, and if you don't plan to use those styles in the whole app you don't need to import it into `App.js`, import it in the component instead

File: src/App.js (hunk 1 above)
Comment (blaise82) [reply src/App.js]: yeah let me do so

**Comment by blaise82** [comment]
> Add `package-lock` file to gitignore, and if you don't plan to use those styles in the whole app you don't need to import it into `App.js`, import it in the component instead

from my point of view, I don't think it's a good thing to put the package-lock file gitignore file.

File: src/redux/actions/loginAction.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
+import axios from 'axios';
+import { FETCH_USERS_REQUEST, FETCH_USERS_SUCCESS, FETCH_USERS_FAILURE, CLOSE_MESSAGE } from '../types/loginTypes';
```
Comment (hakoemmy) [inline src/redux/actions/loginAction.js]: @blaise82 Could you use meaningful action types? Are you really fetching users? consider my suggestions below:
```suggestion
import { LOGIN_REQUEST, LOGIN_SUCCESS, LOGIN_FAILURE, CLOSE_MESSAGE } from '../types/loginTypes';
```

File: src/redux/actions/loginAction.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+				dispatch(fetchUsersSuccess(user));
+			})
+			.catch((error) => {
+				// error.message is the error message
+				dispatch(fetchUsersFailure(error.response.data.error));
+			});
+	};
+};
+
+export const fetchUsersRequest = () => {
```
Comment (hakoemmy) [inline src/redux/actions/loginAction.js]: @blaise82 Even these names of actions creators, tweak them in the way you want to make a sense
```suggestion
export const loginRequest = () => {
```

File: src/redux/actions/loginAction.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+	return {
+		type: FETCH_USERS_REQUEST
+	};
+};
+
+export const fetchUsersSuccess = (user) => {
```
Comment (hakoemmy) [inline src/redux/actions/loginAction.js]: Tweak the as well, an example below:
```suggestion
export const loginSuccess = (user) => {
```

File: src/redux/actions/loginAction.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+	return {
+		type: FETCH_USERS_SUCCESS,
+		payload: user
+	};
+};
+
+export const fetchUsersFailure = (error) => {
```
Comment (hakoemmy) [inline src/redux/actions/loginAction.js]: Please do tweak actions creatores to make a sense.
```suggestion
export const loginFailure = (error) => {
```

File: src/components/LoginForm.js [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+		this.setState({
+			[e.target.name]: e.target.value
+		});
+	};
+	render() {
+		return (
+			<div>
+				<div className={styles.forms}>
+					<div style={{ display: this.props.loginSate.loading }}>
+						<LinearProgress />
```
Comment (hakoemmy) [inline src/components/LoginForm.js:77]: @blaise82  can you consider redirecting a user for a successful login attempt(Say redirect to update profile, even though the page is not there, please do include such route)? Please do include it in your Login Component .

**Review by hakoemmy — COMMENTED** [review]
@blaise82  I left some comments. do walk through them.

**Review by AbdoulNuru — APPROVED** [review]

File: src/__tests__/login.spec.js [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+	</Provider>
+);
+describe('LOGIN', () => {
+	afterEach(cleanup);
+	it('should render login component', () => {
+		const { asFragment } = render(
+			<Provider store={store}>
+				<Login />
+			</Provider>
+		);
```
Comment (AJAkimana) [inline src/__tests__/login.spec.js]: Use the `Component` variable as you have declared above

**Review by hakoemmy — APPROVED** [review]

File: src/__tests__/login.spec.js [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+import { render, fireEvent, cleanup, waitFor } from '@testing-library/react';
+import Login from '../views/Login.View';
+import { Provider } from 'react-redux';
+import store from '../redux/store';
+import userEvent from '@testing-library/user-event';
+
+const Component = render(
+	<Provider store={store}>
+		<Login />
+	</Provider>
```
Comment (AJAkimana) [inline src/__tests__/login.spec.js]: Name this variable according to its role. For instance `loginComponent` or another significant name

File: src/__tests__/login.spec.js [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+				<Login />
+			</Provider>
+		);
+		const email = getByLabelText('email');
+		const password = getByLabelText('password');
+		// const submit = getByLabelText('submit');
+		const form = container.querySelector('form');
+
+		userEvent.type(email, 'email@gmail.com');
+		userEvent.type(password, '');
```
Comment (AJAkimana) [inline src/__tests__/login.spec.js]: Identical block

File: src/__tests__/login.spec.js [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+			<Provider store={store}>
+				<Login />
+			</Provider>
+		);
+		const email = getByLabelText('email');
+		const password = getByLabelText('password');
+		// const submit = getByLabelText('submit');
+		const form = container.querySelector('form');
+		userEvent.type(password, 'password');
+		await form.dispatchEvent(new Event('submit'));
```
Comment (AJAkimana) [inline src/__tests__/login.spec.js]: Identical block

File: src/styles/styles.scss [hunk 12]
```diff
@@ -1,6 +1,7 @@
...
+    font-family: 'Open Sans', sans-serif;
+    margin: 0;
+    padding: 0;
+    overflow-y: hidden;
 
-.center {
-    color: $example-color;
-    text-align: center;
-}
+}
```
Comment (AJAkimana) [inline src/styles/styles.scss]: Add new line

**Review by AJAkimana — COMMENTED** [review]
Refactor this code in the way suggested in the inline comments

**Review by Moise1 — APPROVED** [review]

File: src/__tests__/login.spec.js (hunk 8 above)
Comment (blaise82) [reply src/__tests__/login.spec.js]: Changed

**Review by AJAkimana — COMMENTED** [review]
Remove the  `logger` middleware. It is adding nothing as long as `redux-devtools-extension` installed

**Review by AJAkimana — COMMENTED** [review]
Set the store to remember the token after sign in, in case you refresh a page, closed it or token expired. And lastly, remove all comments that mean nothing

**Pull request merged**